### PR TITLE
Assert on calling main more than once

### PR DESCRIPTION
**Pull request by kripken**
If `callMain` is called more than once, bad things can happen, such as
the stack allocation of argv perhaps overflowing, or issues with the runtime
initialization/shutdown. In general, in the C world it is simply not valid to
call main multiple times, so it seems reasonable to enforce this.

However, there might be a use case I am missing?

Background: #14312

### CODE REVIEW

File: src/postamble.js [hunk 1]
```diff
@@ -119,6 +119,7 @@ function callMain(args) {
 #if ASSERTIONS
   assert(runDependencies == 0, 'cannot call main when async dependencies remain! (listen on Module["onRuntimeInitialized"])');
   assert(__ATPRERUN__.length == 0, 'cannot call main when preRun functions remain to be called');
+  assert(!calledMain, 'in the C world, main() is special in that it should only be called once');
```
Comment (sbc100) [inline src/postamble.js:122]: Maybe mention `callMain` itself as the thing that can't be repeatedly called?

Ironically, although its not advised, calling `_main` directly does't have any of the same issues.

Should we link to an FAQ entry that can help folks who want the semantics of a multi-call main?  What should they do instead?

Comment (kripken) [reply src/postamble.js:122]: Perhaps an FAQ entry is better than this, yeah. Could mention MODULARIZE etc.

Or, perhaps we could try to fix callMain to be runnable multiple times.

Comment (sbc100) [reply src/postamble.js:122]: Yes I guess there are two solutions we would want to recommend:

1. Use `MODULARIZE` and instantiate a new module for each call
2. Convert to a library that exposes the same functionality as `main` under a different name which can be multi-called.

**Comment by vadimkantorov** [comment]
As you know, I've been calling main more than once without much problems - it saves a lot of time for module reinit. https://github.com/emscripten-core/emscripten/issues/12219

In fact, memory reset + consistent file handle closure works well (and I manually do callMain several times to call multiple programs): https://github.com/busytex/busytex/blob/main/busytex_pipeline.js#L302

I hope emscripten doesn't block this scenario. For heavy programs like TexLive toolchain compiled into a single binary, it's important to be as fast as possible

File: src/postamble.js (hunk 1 above)
Comment (vadimkantorov) [reply src/postamble.js:122]: There is also a problem with hard-to-obtain exit code

**Comment by sbc100** [comment]
> As you know, I've been calling main more than once without much problems - it saves a lot of time for module reinit. #12219
> 
> In fact, memory reset + consistent file handle closure works well (and I manually do callMain several times to call multiple programs): https://github.com/busytex/busytex/blob/main/busytex_pipeline.js#L302
> 
> I hope emscripten doesn't block this scenario. For heavy programs like TexLive toolchain compiled into a single binary, it's important to be as fast as possible

**Comment by sbc100** [comment]
> As you know, I've been calling main more than once without much problems - it saves a lot of time for module reinit. #12219
> 
> In fact, memory reset + consistent file handle closure works well (and I manually do callMain several times to call multiple programs): https://github.com/busytex/busytex/blob/main/busytex_pipeline.js#L302
> 
> I hope emscripten doesn't block this scenario. For heavy programs like TexLive toolchain compiled into a single binary, it's important to be as fast as possible

I think the intention it to make new users aware of potentially dangerous patterns, so only those (like yourself) who really understand the consequences will end up doing such things.

If this lands I imagine you will just need to add `calledMain = false` to list of things you due during your "state reset" code.    This seems fairly reasonable to me.   The "state reset" code that you maintain is likely to require such changes over time as emscripten evolves since it is very much tinkering with internal stuff.

**Comment by vadimkantorov** [comment]
And part of the problem is that re-init is relatively slow, and that people who rely on calling the module multiple times would likely have the same problem. So supporting some pattern of state reset may be a usefule feature for relatively large number of users.

**Comment by sbc100** [comment]
> And part of the problem is that re-init is relatively slow, and that people who rely on calling the module multiple times would likely have the same problem. So supporting some pattern of state reset may be a usefule feature for relatively large number of users.

I think its pretty common that no state reset at all is needed.  Many programs have main functions that are more or less idempotent.

**Comment by sbc100** [comment]
> > And part of the problem is that re-init is relatively slow, and that people who rely on calling the module multiple times would likely have the same problem. So supporting some pattern of state reset may be a usefule feature for relatively large number of users.
> 
> I think its pretty common that no state reset at all is needed. Many programs have main functions that are more or less idempotent.

In this case the startup code (static ctors and whatnot) will just run once.. and then main is called more like a library function.  For ffpmeg this is apparently enough.

**Comment by sbc100** [comment]
> > > And part of the problem is that re-init is relatively slow, and that people who rely on calling the module multiple times would likely have the same problem. So supporting some pattern of state reset may be a usefule feature for relatively large number of users.
> > 
> > 
> > I think its pretty common that no state reset at all is needed. Many programs have main functions that are more or less idempotent.
> 
> In this case the startup code (static ctors and whatnot) will just run once.. and then main is called more like a library function. For ffmpeg (for example) this is apparently enough.

**Comment by vadimkantorov** [comment]
Also that. For me it was working fine for a long time, until one of TexLive programs relied on some static variables in some bad way. So memory reset was needed.

**Comment by vadimkantorov** [comment]
For the scenario of calling main multiple times, it'd better return exit code :) (my old problem)

**Comment by kripken** [comment]
Let's close this, as I wouldn't want to block advanced users doing potentially unsafe things that they know are ok to do. Adding a flag to disable this assertion could avoid that too, but it adds more complexity, and I'm not sure it's worth it - user reports on problems with this are extremely rare.

**Pull request closed without merging**